lib/tools/src/hprof.erl

```diff
@@ -466,11 +471,8 @@ format_out(Device, Fmt, Args) ->
 enable_pattern('_', '_', '_', _Acc) ->
     %% need to re-trace everything, probably some new modules were loaded
     %% discard any existing trace pattern
-    lists:foldl(
-        fun({Mod, _}, {Total, Acc}) ->
-            Plus = erlang:trace_pattern({Mod, '_', '_'}, true, [call_memory]),
-            {Total + Plus, Acc#{Mod => Mod:module_info(functions)}}
-        end, {0, #{}}, code:all_loaded());
+    erlang:trace_pattern(on_load, true, [call_memory]),
+    {erlang:trace_pattern({'_', '_', '_'}, true, [call_memory]), #{ '_' => undefined }};
```

Comment:
I suspect that having `#{'_' => undefined}` is correct. It is used in this line: 
https://github.com/erlang/otp/blob/master/lib/tools/src/hprof.erl#L280C1-L280C85

So the public API `get_trace_map` will be broken.

I haven't tested this change, but I think it will also break the flow when patterns are added/removed (with the gen_server keeping Map in state).

I understand that due to various race conditions it's possible to have more (or less) modules traced when tracing is enabled/disabled while code loading is in progress. But to me, this isn't a big deal.
```

```

Reply:
> So the public API get_trace_map will be broken.

Good catch, I will fix that.

> I understand that due to various race conditions it's possible to have more (or less) modules traced when tracing is enabled/disabled while code loading is in progress. But to me, this isn't a big deal.

If I run `hprof:profile(fun() -> sofs:relation([{a,b}]) end)` in a fresh the shell, I will get no profiling information about the sofs module, to me that is surprising.

Reply:
Right, the interactively loaded modules are unknown to `hprof` server process, and therefore don't show up in the profile. It is a complex problem, if modules may appear and disappear at any point in time. To be precise, I think `hprof` may even exhibit crashes or other unexpected behaviour if modules are unloaded. Probably worth a note in the documentation.

Given that there is no way to subscribe to "new module loaded" event, `hprof` in it's current share would require a separate code path for "trace all modules" use-case. E.g. `trace_map`: https://github.com/erlang/otp/blob/master/lib/tools/src/hprof.erl#L265 may become something like `trace_map() | all`.

Reply:
Pushed update that makes `trace_map() :: #{ module() => [{atom(),arity}] } | all` and updated the code to respect that in all cases.

As you said, a problematic situation arrives when you do a trace on all and the disable trace in some places. I've decided that for now the `get_trace_map()` will still show it as `all` but the disabled MFA will not be part of collect_data. The only way to solve it would be to keep a "disabled" map and then substract that from `erlang:all_loaded()` when `get_trace_map` is called.